homeassistant/components/tuya/alarm_control_panel.py

```diff
@@ -0,0 +1,149 @@
+    # Alarm
+    # https://developer.tuya.com/en/docs/iot/categorymal?id=Kaiuz33clqxaf
+    "mal": (
+        TuyaAlarmEntityDescription(
+            key="master_mode",
+            name="Alarm",
+            battery_percentage=DPCode.BATTERY_PERCENTAGE,
+        ),
+    ),
+}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up Tuya alarm dynamically through Tuya discovery."""
+    hass_data: HomeAssistantTuyaData = hass.data[DOMAIN][entry.entry_id]
+    _LOGGER.info(hass_data)
+
+    @callback
+    def async_discover_device(device_ids: list[str]) -> None:
+        """Discover and add a discovered Tuya siren."""
+        entities: list[TuyaAlarmEntity] = []
+        for device_id in device_ids:
+            device = hass_data.device_manager.device_map[device_id]
+            if descriptions := ALARM.get(device.category):
+                for description in descriptions:
+                    if (
+                        description.key in device.function
+                        or description.key in device.status
+                    ):
+                        entities.append(
+                            TuyaAlarmEntity(
+                                device, hass_data.device_manager, description
+                            )
+                        )
+        async_add_entities(entities)
+
+    async_discover_device([*hass_data.device_manager.device_map])
+
+    entry.async_on_unload(
+        async_dispatcher_connect(hass, TUYA_DISCOVERY_NEW, async_discover_device)
+    )
+
+
+class TuyaAlarmEntity(TuyaEntity, AlarmControlPanelEntity):
+    """Tuya Alarm Entity."""
+
+    def __init__(
+        self,
+        device: TuyaDevice,
+        device_manager: TuyaDeviceManager,
+        description: TuyaAlarmEntityDescription,
+    ) -> None:
+        """Init Tuya Alarm."""
+        super().__init__(device, device_manager)
+        self.entity_description = description
+        self._attr_unique_id = f"{super().unique_id}{description.key}"
+
+    @property
+    def state(self):
+        """Return the current Alarm State."""
+        return self.device.status.get(DPCode.ALARM_STATE)
+
+    @property
+    def code_format(self) -> str | None:
+        """Regex for code format or None if no code is required."""
+        return self._attr_code_format
+
+    @property
+    def changed_by(self) -> str | None:
+        """Last change triggered by."""
+        return self._attr_changed_by
+
+    @property
+    def code_arm_required(self) -> bool:
+        """Whether the code is required for arm actions."""
+        return self._attr_code_arm_required
+
+    @property
+    def supported_features(self) -> int:
+        """Return the list of supported features."""
+        return (
+            SUPPORT_ALARM_ARM_HOME
+            | SUPPORT_ALARM_ARM_AWAY
+            | SUPPORT_ALARM_ARM_NIGHT
+            | SUPPORT_ALARM_TRIGGER
+        )
```

Comment:
We need to make sure the value constraint from the API actually indicates support for these before setting them.

Additionally, I don't see night arming implemented?